# Aim : To Check the Relu Activation Function with Weight initializer: he_normal

In [1]:
# TensorFlow and tf.keras
import tensorflow as tf

# Helper libraries
import numpy as np
from tensorflow.keras import initializers
from tensorflow.python.keras import activations

print(tf.__version__)

# downloading fashion_mnist data
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images / 255.0

test_images = test_images / 255.0              


import mlflow.tensorflow
mlflow.tensorflow.autolog()

activation = tf.keras.activations.relu
initializer = tf.keras.initializers.he_normal

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

2.5.0


2021/06/13 15:19:35 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of tensorflow. If you encounter errors during autologging, try upgrading / downgrading tensorflow to a supported version, or try upgrading MLflow.


Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


In [2]:
# Checking weight initializer used 
print(model.layers[1].kernel_initializer)
print(model.layers[2].kernel_initializer)

In [3]:
with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Epoch 1/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.5015 - accuracy: 0.8245
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3768 - accuracy: 0.8636
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3387 - accuracy: 0.8769
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3164 - accuracy: 0.8834
Epoch 5/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2976 - accuracy: 0.8900
Epoch 6/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2820 - accuracy: 0.8955
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2707 - accuracy: 0.9003
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2611 - accuracy: 0.9028
Epoch 9/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2500 - accuracy: 0.9076
Epoch 10/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.243

### Observations:
- For epoch = 10 we get `train_loss = 0.24`, `train_acc = 0.90`, `test_loss = 0.33`, `test_acc= 0.88`